### PR TITLE
TEAMS: Added support for Application based Authentication 

### PR DESCRIPTION
**Pull request by NikCharlebois**
#### Pull Request (PR) description
* Added support for ApplicationId, tenantId and CertificateThumbprint across Teams resources.

#### This Pull Request (PR) fixes the following issues
* N/A

### CODE REVIEW

File: docs/docs/user-guide/get-started/authentication-and-permissions.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ The following table provides an overview of what authentication methods are supp
 | *Planner** | Microsoft.Graph.Authentication<br /> (Connect-MgGraph) | ![Check](../../Images/check.png) | ![Cross](../../Images/cross.png) | ![Cross](../../Images/cross.png) | ![Cross](../../Images/cross.png) | ![Cross](../../Images/cross.png) |
```
Comment (ykuijs) [inline docs/docs/user-guide/get-started/authentication-and-permissions.md:31]: Shouldn't Planner support be added as well?

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.schema.mof [hunk 2]
```diff
@@ -4,6 +4,9 @@ class MSFT_TeamsUpgradeConfiguration : OMI_BaseResource
...
     [Write, Description("The SfBMeetingJoinUx property allows admins to specify which app is used to join Skype for Business meetings, even after the user has been upgraded to Teams. Allowed values are: 'SkypeMeetingsApp' and 'NativeLimitedClient'. 'NativeLimitedClient' means the existing Skype for Business rich client will be used, but since the user is upgraded, only meeting functionality is available. Calling and Messaging are done via Teams. 'SkypeMeetingsApp' means use the web-downloadable app. This setting can be useful for organizations that have upgraded to Teams and no longer want to install Skype for Business on their users' computers."),ValueMap{"SkypeMeetingsApp","NativeLimitedClient"}, Values{"SkypeMeetingsApp","NativeLimitedClient"}] String SfBMeetingJoinUx;
-    [Required, Description("Credentials of the Teams Admin"), EmbeddedInstance("MSFT_Credential")] string Credential;
+    [Write, Description("Credentials of the Teams Admin"), EmbeddedInstance("MSFT_Credential")] string Credential;
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.schema.mof:7]: Please check the various Readme files for consistency. Some say "Teams Global Admin", this one says "Teams Admin".

I think it makes most sense to use "Teams Admin".

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.psm1 [hunk 3]
```diff
@@ -167,7 +209,7 @@ function Test-TargetResource
     Write-Verbose -Message "Target Values: $(Convert-M365DscHashtableToString -Hashtable $PSBoundParameters)"
 
     $ValuesToCheck = $PSBoundParameters
-    $ValuesToCheck.Remove('Credential') | Out-Null
+
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.psm1:212]: (Picky mode) Please remove one empty line.

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsMeetingBroadcastConfiguration/MSFT_TeamsMeetingBroadcastConfiguration.psm1 [hunk 4]
```diff
@@ -266,6 +320,9 @@ function Export-TargetResource
         $params = @{
             Identity   = "Global"
             Credential = $Credential
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_TeamsMeetingBroadcastConfiguration/MSFT_TeamsMeetingBroadcastConfiguration.psm1:322]: Please align the =-signs in the hashtable.

**Review by ykuijs — COMMENTED** [review]
Few minor comments

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.schema.mof (hunk 2 above)
Comment (NikCharlebois) [reply Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.schema.mof:7]: Done. Changed "Teams Global Admin" to "Teams Admin" across the board.

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.psm1 (hunk 3 above)
Comment (NikCharlebois) [reply Modules/Microsoft365DSC/DSCResources/MSFT_TeamsUpgradeConfiguration/MSFT_TeamsUpgradeConfiguration.psm1:212]: Done

File: Modules/Microsoft365DSC/DSCResources/MSFT_TeamsMeetingBroadcastConfiguration/MSFT_TeamsMeetingBroadcastConfiguration.psm1 (hunk 4 above)
Comment (NikCharlebois) [reply Modules/Microsoft365DSC/DSCResources/MSFT_TeamsMeetingBroadcastConfiguration/MSFT_TeamsMeetingBroadcastConfiguration.psm1:322]: Done

File: Modules/Microsoft365DSC/DSCResources/MSFT_PlannerPlan/MSFT_PlannerPlan.psm1 [hunk 5]
```diff
@@ -120,10 +140,15 @@ function Get-TargetResource
...
-                Ensure     = 'Present'
-                Credential = $Credential
+                Title                 = $Title
+                OwnerGroup            = $OwnerGroupValue
+                Ensure                = 'Present'
+                Credential            = $Credential
+                ApplicationId         = $ApplicationId
+                TenantId              = $TenantId
+                CertificateThumbprint = $CertificateThumbprint
+		        ApplicationSecret     = $ApplicationSecret
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_PlannerPlan/MSFT_PlannerPlan.psm1:150]: Please fix alignment

File: Modules/Microsoft365DSC/DSCResources/MSFT_PlannerTask/MSFT_PlannerTask.psm1 [hunk 6]
```diff
@@ -155,22 +175,27 @@ function Get-TargetResource
...
+                PercentComplete       = $task.PercentComplete
+                StartDateTime         = $StartDateTimeValue
+                DueDateTime           = $DueDateTimeValue
+                Notes                 = $NotesValue
+                Ensure                = "Present"
+                Credential            = $Credential
+                ApplicationId         = $ApplicationId
+                TenantId              = $TenantId
+                CertificateThumbprint = $CertificateThumbprint
+		        ApplicationSecret     = $ApplicationSecret
```
Comment (ykuijs) [inline Modules/Microsoft365DSC/DSCResources/MSFT_PlannerTask/MSFT_PlannerTask.psm1:197]: Please fix alignment

**Review by ykuijs — CHANGES_REQUESTED** [review]
Almost there!

File: Modules/Microsoft365DSC/DSCResources/MSFT_PlannerPlan/MSFT_PlannerPlan.psm1 (hunk 5 above)
Comment (NikCharlebois) [reply Modules/Microsoft365DSC/DSCResources/MSFT_PlannerPlan/MSFT_PlannerPlan.psm1:150]: Done

File: Modules/Microsoft365DSC/DSCResources/MSFT_PlannerBucket/MSFT_PlannerBucket.schema.mof [hunk 7]
```diff
@@ -5,6 +5,11 @@ class MSFT_PlannerBucket : OMI_BaseResource
...
     [Write, Description("Id of the Bucket, if known.")] string BucketId;
     [Write, Description("Present ensures the Plan exists, absent ensures it is removed"), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
-    [Required, Description("Credentials of the account to authenticate with."), EmbeddedInstance("MSFT_Credential")] string Credential;
+    [Write, Description("Credentials of the account to authenticate with."), EmbeddedInstance("MSFT_Credential")] string Credential;
+    [Write, Description("Id of the Azure Active Directory application to authenticate with.")] String ApplicationId;
+    [Write, Description("Id of the Azure Active Directory tenant used for authentication.")] String TenantId;
+    [Write, Description("Secret of the Azure Active Directory tenant used for authentication."), EmbeddedInstance("MSFT_Credential")] String ApplicationSecret;
+    [Write, Description("Thumbprint of the Azure Active Directory application's authentication certificate to use for authentication.")] String CertificateThumbprint;
+    [Write, Description("Managed ID being used for authentication.")] Boolean ManagedIdentity;
```
[the remark this answers is not in the document]
Comment (NikCharlebois) [reply Modules/Microsoft365DSC/DSCResources/MSFT_PlannerBucket/MSFT_PlannerBucket.schema.mof:13]: Will need to align across all resources. Will submit a separate PR.

**Review by ykuijs — APPROVED** [review]
LGTM

**Pull request merged**